Frameworks/CoreGraphics/CGFont.mm

```diff
@@ -753,46 +323,54 @@ CGFloat CGFontGetStemV(CGFontRef font) {
  @Status Stub
  @Notes
 */
-CFTypeID CGFontGetTypeID() {
+CFStringRef CGFontCopyGlyphNameForGlyph(CGFontRef font, CGGlyph glyph) {
     UNIMPLEMENTED();
     return StubReturn();
 }
 
-// TODO: functions below are not part of offical exports, but they are also exported
-// to be used by other framework components, we should consider moving them to a shared library
-unsigned long CGFontGetGlyphsForUnichars(id font, unsigned short* chars, unsigned short* glyphs, int num) {
-    CGFontGetGlyphs(font, chars, num, glyphs);
-    return TRUE;
+/**
+ @Status Interoperable
+ @Notes
+*/
+size_t CGFontGetNumberOfGlyphs(CGFontRef font) {
+    return font ? font->_dwriteFontFace->GetGlyphCount() : 0;
 }
 
-float CGFontGetDescentAtSize(id font, float size) {
-    _CGFontLock();
-    FT_Face face = (FT_Face)(DWORD)[font _sizingFontHandle];
-    CGFontSetFTFontSize(font, face, size);
-
-    float ret = ((float)face->size->metrics.descender) / 64.0f;
-    _CGFontUnlock();
-
-    return ret;
-}
+/**
+ @Status Interoperable
+*/
+bool CGFontGetGlyphAdvances(CGFontRef font, const CGGlyph* glyphs, size_t count, int* advances) {
+    if (!font) {
+        return false;
+    }
 
-float CGFontGetAscentAtSize(id font, float size) {
-    _CGFontLock();
-    FT_Face face = (FT_Face)(DWORD)[font _sizingFontHandle];
-    CGFontSetFTFontSize(font, face, size);
+    ComPtr<IDWriteFontFace1> fontFace1;
+    font->_dwriteFontFace.As(&fontFace1);
 
-    float ret = ((float)face->size->metrics.ascender) / 64.0f;
-    _CGFontUnlock();
+    if (FAILED(fontFace1->GetDesignGlyphAdvances(count, glyphs, advances, false))) {
+        TraceError(g_logTag, L"Unable to get glyph advances");
+        return false;
+    }
 
-    return ret;
+    return true;
 }
 
-static pthread_mutex_t _fontLock = PTHREAD_MUTEX_INITIALIZER;
-
-void _CGFontLock() {
-    pthread_mutex_lock(&_fontLock);
+/**
+ @Status Interoperable
+*/
+int CGFontGetUnitsPerEm(CGFontRef font) {
+    return font ? __CGFontGetDWriteMetrics(font).designUnitsPerEm : 0;
 }
 
-void _CGFontUnlock() {
-    pthread_mutex_unlock(&_fontLock);
-}
+/**
+ @Status Interoperable
+ @Notes
+*/
+CFTypeID CGFontGetTypeID() {
+    static dispatch_once_t initOnce = 0;
+    dispatch_once(&initOnce,
+                  ^{
+                      __kCGFontTypeID = _CFRuntimeRegisterClass(&__CGFontClass);
+                  });
+    return __kCGFontTypeID;
+}
```

Comment:
magic statics will work a treat here, but there's little wrong with dispatch_once

---
In reply to: [83340826](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r83340826) [](ancestors = 83340826)